File: Checker/LambdaP.hs
module LambdaP where

type Name = String
data Kind = Kstar 
          | Kpi Name Type Kind
            
data Type = Tvar Name 
          | Tall Name Type Type 
          | Tapp Type Term
            
data Term = Mvar Name
          | Mapp Term Term 
          | Mabs Name Type Term
            
dummy = "_"
karr = Kpi dummy 
tarr = Tall dummy

instance Num Type where
  fromInteger = Tvar . show
  (+) = undefined
  (*) = undefined
  abs = undefined
  signum = undefined
  
k0 = Kstar
k1 = karr 0 k0
t1 = tarr 0 0
m1 = Mabs "x" 0 (Mvar "x")

instance Show Kind where            
  show Kstar = "*"
  show (Kpi n t k) = concat["(",n,":",show t,")->",show k]
  
instance Show Type where
  show (Tvar n) = n
  show (Tall n t1 t2) = concat ["(∀",n,":",show t1,")",show t2]
  show (Tapp t1 t2) = unwords [show t1,show t2]
  
instance Show Term where
  show (Mvar n) = n
  show (Mapp t1 t2) = unwords [show t1,show t2] 
  show (Mabs n t1 t2) = concat ["(","λ",n,":",show t1,".",show t2,")"]

  

File: Checker/lambdap.ml
type name = string

type kind = Kstar 
          | Kpi of name * typ * kind
and typ = Tvar of name
        | Tall of name * typ * typ
        | Tapp of typ * term

and term = Mvar of name
         | Mapp of term * term
         | Mabs of name * typ * term

let dummy = "_"
let karr (x,y) = Kpi (dummy,x,y)
let tarr (x,y) = Tall (dummy,x,y)

let k0 = Kstar
let t0 = Tvar "0"
let k1 = karr (t0,k0)
let m1 = Mabs ("x",t0,Mvar "x")
